From 5ef58789d06f6221a27ba0ec66663bfa726a0fee Mon Sep 17 00:00:00 2001
From: Cezary Rojewski <cezary.rojewski@intel.com>
Date: Mon, 3 Sep 2018 14:42:21 +0200
Subject: [PATCH 189/292] ASoC: Intel: 4.19-rc1 and 4.20-rc1 rebase fixups.

This is a temporary patch.

4.19-rc1:
- Fix soc_dai_hw_params NULL dereference.

This patch fixes NULL dereference panic within soc_dai_hw_params
introduced by latest soc-framework changes. CoE does not instantiate the
snd_soc_pcm_runtine rtd field for pcm substreams.

4.20-rc1:
- Fix hw_params for skl tplg.

CODEC-CODEC dais bind routines have been removed and merged into normal
dais bind methods. Because of this, the amount of paths available for
source and sink widgets has also changed.
skl_tplg_be_set_<>_pipe params can no longer use the default -EIO
return value since it is no longer guaranteed that all paths contain at
least one skl widget.

- Fix general segmentation fault for snd_soc_register_card.

The initial implementation for multiplatform support found in soc-core.c
enforced existence of platform for any dai_link. If machine driver
instantiation fails, e.g.: due to codec driver being not yet available,
the probe is deferred and the current card resources are freed.
Despite being freed, platform references are not NULLed.
Deferred probe will attempt to access that data causing unexpected
behaviour.

Change-Id: I06600e84714863ec37f551ed5373830dec10c29c
Signed-off-by: Cezary Rojewski <cezary.rojewski@intel.com>
---
 sound/soc/intel/skylake/skl-topology.c |  4 ++--
 sound/soc/intel/skylake/skl-topology.h |  2 +-
 sound/soc/soc-core.c                   | 15 +++++++++++++--
 sound/soc/soc-pcm.c                    |  2 +-
 4 files changed, 17 insertions(+), 6 deletions(-)

diff --git a/sound/soc/intel/skylake/skl-topology.c b/sound/soc/intel/skylake/skl-topology.c
index 6ba8311b56f8..cd292f650bbc 100644
--- a/sound/soc/intel/skylake/skl-topology.c
+++ b/sound/soc/intel/skylake/skl-topology.c
@@ -2960,7 +2960,7 @@ static int skl_tplg_be_set_src_pipe_params(struct snd_soc_dai *dai,
 				struct skl_pipe_params *params)
 {
 	struct snd_soc_dapm_path *p;
-	int ret = -EIO;
+	int ret = 0;
 
 	snd_soc_dapm_widget_for_each_source_path(w, p) {
 		if (p->connect && is_skl_dsp_widget_type(p->source, dai->dev) &&
@@ -2985,7 +2985,7 @@ static int skl_tplg_be_set_sink_pipe_params(struct snd_soc_dai *dai,
 	struct snd_soc_dapm_widget *w, struct skl_pipe_params *params)
 {
 	struct snd_soc_dapm_path *p = NULL;
-	int ret = -EIO;
+	int ret = 0;
 
 	snd_soc_dapm_widget_for_each_sink_path(w, p) {
 		if (p->connect && is_skl_dsp_widget_type(p->sink, dai->dev) &&
diff --git a/sound/soc/intel/skylake/skl-topology.h b/sound/soc/intel/skylake/skl-topology.h
index 6850f447c0da..dfbc0eb36301 100644
--- a/sound/soc/intel/skylake/skl-topology.h
+++ b/sound/soc/intel/skylake/skl-topology.h
@@ -537,7 +537,7 @@ int skl_dsp_set_dma_control(struct skl_sst *ctx, u32 *caps,
 void skl_tplg_set_be_dmic_config(struct snd_soc_dai *dai,
 	struct skl_pipe_params *params, int stream);
 int skl_tplg_init(struct snd_soc_component *component,
-				struct hdac_bus *ebus);
+				struct hdac_bus *bus);
 struct skl_module_cfg *skl_tplg_fe_get_cpr_module(
 		struct snd_soc_dai *dai, int stream);
 int skl_tplg_update_pipe_params(struct device *dev,
diff --git a/sound/soc/soc-core.c b/sound/soc/soc-core.c
index 50617db05c46..3814f9359794 100644
--- a/sound/soc/soc-core.c
+++ b/sound/soc/soc-core.c
@@ -2771,7 +2771,7 @@ int snd_soc_register_card(struct snd_soc_card *card)
 			dev_err(card->dev, "ASoC: failed to init link %s\n",
 				link->name);
 			mutex_unlock(&client_mutex);
-			return ret;
+			goto err;
 		}
 	}
 	mutex_unlock(&client_mutex);
@@ -2791,7 +2791,18 @@ int snd_soc_register_card(struct snd_soc_card *card)
 	mutex_init(&card->mutex);
 	mutex_init(&card->dapm_mutex);
 
-	return snd_soc_bind_card(card);
+	ret = snd_soc_bind_card(card);
+	if (ret)
+		goto err;
+
+	return 0;
+
+err:
+	for_each_card_prelinks(card, i, link)
+		if (link->platform)
+			link->platform = NULL;
+
+	return ret;
 }
 EXPORT_SYMBOL_GPL(snd_soc_register_card);
 
diff --git a/sound/soc/soc-pcm.c b/sound/soc/soc-pcm.c
index 9a0484779a46..32ba43c53205 100644
--- a/sound/soc/soc-pcm.c
+++ b/sound/soc/soc-pcm.c
@@ -860,7 +860,7 @@ int soc_dai_hw_params(struct snd_pcm_substream *substream,
 	int ret;
 
 	/* perform any topology hw_params fixups before DAI  */
-	if (rtd->dai_link->be_hw_params_fixup) {
+	if (rtd && rtd->dai_link->be_hw_params_fixup) {
 		ret = rtd->dai_link->be_hw_params_fixup(rtd, params);
 		if (ret < 0) {
 			dev_err(rtd->dev,
-- 
2.21.0

